Toggleable product variant

mmenozzi opened the pull request:
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | implement #11115 
| License         | MIT

With this PR we add an enabled/disabled status (toggleable) to the product variant as described in #11115.

stloyd commented on app/migrations/Version20200309172908.php:
```diff
@@ -0,0 +1,35 @@
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->abortIf($this->connection->getDatabasePlatform()->getName() !== 'mysql', 'Migration can only be executed safely on \'mysql\'.');
+
+        $this->addSql('ALTER TABLE sylius_product_variant ADD enabled TINYINT(1) NOT NULL');
```
This should have `DEFAULT 1` if you force it to not be `NULL` cause in tables that contain data it will fail IIRC.
```suggestion
        $this->addSql('ALTER TABLE sylius_product_variant ADD enabled TINYINT(1) DEFAULT 1 NOT NULL');
```

mmenozzi replied:
Yeah you're right we just copied what has been done for other `enabled` columns in the core but to avoid BC break we agree that the default value should be provided here. We just committed your suggestion.

mmenozzi replied:
Sorry but I just realized that this needs column definition in Doctrine mapping too. And as stated here https://stackoverflow.com/questions/3376881/default-value-in-doctrine Doctrine column definition mapping isn't portable across various DBMS.

The default value of 1 (true) is already set at PHP object level so I don't know if it's really needed.

@pamil what do you think?

stloyd replied:
Second option always used in legacy systems is:
- add field without `NOT NULL`,
- set default value on all items in that table where value `IS NULL`,
- add `ALTER TABLE` with proper `NOT NULL`,
So you need 3 queries, but you avoid adding custom config into field settings

mmenozzi replied:
What if we simply add the following query to the migration (after the `ALTER TABLE`)?

```sql
$this->addSql('UPDATE sylius_product_variant SET enabled=1');
```

stloyd replied:
It will not work with setting value that was forced to be set by `NOT NULL`, it will fail already on that query as mentioned.
```suggestion
        $this->addSql('ALTER TABLE sylius_product_variant ADD enabled TINYINT(1) NULL');
        $this->addSql('UPDATE sylius_product_variant SET enabled = 1');
        $this->addSql('ALTER TABLE sylius_product_variant ADD enabled TINYINT(1) NOT NULL');
```

mmenozzi replied:
Ok I'd also like to have a core team member opinion here.

mmenozzi replied:
Btw it's not true that it doesn't work with existing variants. I tested on MySQL and MariaDB, the statement:

```sql
ALTER TABLE sylius_product_variant ADD enabled TINYINT(1) NOT NULL
```

Will work even with existing variants.

mmenozzi replied:
Added the following statement to the migration:

```sql
UPDATE sylius_product_variant SET enabled = 1
```
I'm going to resolve this.

stloyd commented on src/Sylius/Component/Product/Resolver/DefaultAvailableProductOptionValuesResolverInterface.php:
```diff
@@ -0,0 +1,38 @@
+use Sylius\Component\Product\Model\ProductOptionInterface;
+use Sylius\Component\Product\Model\ProductOptionValueInterface;
+
+class DefaultAvailableProductOptionValuesResolverInterface implements AvailableProductOptionValuesResolverInterface
```
Class named interface?

mmenozzi replied:
Fixed. Thanks.

mmenozzi commented:
Hi guys, this PR is ready for review. Let us know if you prefer to have it squashed into a single commit.

lchrusciel commented on src/Sylius/Component/Product/Resolver/DefaultAvailableProductOptionValuesResolver.php:
```diff
@@ -0,0 +1,39 @@
+                        return true;
+                    }
+                }
+                return false;
```
```suggestion

                return false;
```

mmenozzi commented on features/product/viewing_products/viewing_different_price_for_different_product_variants_selected_with_options.feature:
```diff
@@ -21,10 +21,4 @@ Feature: Viewing different price for different product variants selected with op
         Then I should see the product price "$25.00"
-
-    @ui @javascript
-    Scenario: Viewing an unavailable message if selected option combination is unavailable
```
Because of https://github.com/Sylius/Sylius/pull/11241#discussion_r404822326.
Given that the "match" variant selection method now only show "avaliable" options is not possible anymore to select an unavailable combination.

lchrusciel replied:
What will happen if we will have matrix option 2 x 2, and only diagonal variants will be available? 
E.g
| Size\Color  | Yellow  |  Blue |
|---|---|---|
| S  | X |   |
|  L |   | X |

What will be displayed to the customer in select boxes?

mmenozzi replied:
Both values for both options because there's at least one enabled variant for at least one value for  Yellow, one for Blue, one for S and one for L. Do you think that a Behat scenario is needed to verify that?

lchrusciel replied:
I would say so. Maybe this scenario should be improved, but this would be the exact case when someone would choose small blue from my previous example.

mmenozzi replied:
Ok I'll do it!

mmenozzi replied:
Done: https://github.com/Sylius/Sylius/pull/11241/commits/df192a303e3f674b113df8261cbb787633b8f637

mmenozzi commented on features/product/viewing_products/viewing_different_price_for_different_product_variants.feature:
```diff
@@ -7,8 +7,10 @@ Feature: Viewing different price for different product variants
     Background:
         Given the store operates on a single channel in "United States"
         And the store has a "Wyborowa Vodka" configurable product
+        And the product "Wyborowa Vodka" has "Wyborowa Lemon" variant priced at "$12.55"
```
Yes, I think so. I wanted to be sure that if a first variant of a product is disabled, its price is not used as the default price of the product. Which in turns means that the `DefaultProductVariantResolver` should resolve to first enabled variant. See 2d6f816eeecc42e104476a2867af5efda6a09654.

mmenozzi commented on features/product/viewing_products/viewing_product_enabled_variants_options_only.feature:
```diff
@@ -0,0 +1,18 @@
+    @ui
+    Scenario: Seeing only enabled variants options
+        When I view product "Super Cool T-Shirt"
+        Then I should not be able to select the "Yellow" color option value
```
Yes. We believe that if variants are disabled them should not be shown in shop in any way. Think about the brand who sell the "Super Cool T-Shirt" which decides to discontinue the "Yellow" color. Then all sizes of that color will be disabled and there's no reason to keep showing the "Yellow" color in the color options list.

mmenozzi commented on src/Sylius/Behat/Context/Setup/ProductContext.php:
```diff
@@ -869,6 +869,94 @@ public function productHasVariant(ProductInterface $product, string $productVari
+    }
+
+    /**
+     * @Given /^(this product) "([^"]+)" ([^"]+) variant is disabled$/
```
Probably that method could be removed. I think that in the latest version only the `\Sylius\Behat\Context\Setup\ProductContext::theProductVariantIsDisabled`  is used. I will check it.

mmenozzi replied:
Done: https://github.com/Sylius/Sylius/pull/11241/commits/0396e066dc2a2097c2417df046a08e119ac7b38f

mmenozzi commented on src/Sylius/Behat/Page/Shop/Product/ShowPage.php:
```diff
@@ -269,6 +291,7 @@ protected function getDefinedElements(): array
             'quantity' => '[data-test-quantity]',
             'validation_errors' => '[data-test-cart-validation-error]',
             'variant_radio' => '[data-test-product-variants] tbody tr:contains("%variantName%") input',
+            'variants_rows' => '[data-test-product-variants] tbody tr',
```
I just copied the line above which uses tags as well. But I agree. I'll refactor with something like:
```php
'variants_rows' => '[data-test-product-variants-row]',
```
Adding that data test attribute to the product page variants rows.

lchrusciel replied:
thanks a lot. We didn't catch all of them during the last rework

mmenozzi replied:
Done in https://github.com/Sylius/Sylius/pull/11241/commits/520b3ced2d3b428e5cbc768863503e7e38e3851e

mmenozzi commented on src/Sylius/Bundle/ProductBundle/Form/Type/ProductOptionValueChoiceType.php:
```diff
@@ -13,14 +13,24 @@
+    /** @var AvailableProductOptionValuesResolverInterface */
+    private $availableProductOptionValuesResolver;
+
+    public function __construct(AvailableProductOptionValuesResolverInterface $availableProductOptionValuesResolver)
```
Why is this a BC break?

Anyway maybe we can keep it but with a setter injection. In the end it is used only when the `only_available_values` option is set and we can refactor to fallback on the old behavior if the `$availableProductOptionValuesResolver` property is not set.

lchrusciel replied:
It is a BC Break, because one could redeclare class form type definition (for whatever reason) and with this change, the container can't be built. We just need to allow for null here and check if the argument is set before the execution of new logic.

mmenozzi replied:
Ok so if I refactor with a setter injection, allowing the `$availableProductOptionValuesResolver` to be null, will you merge it sooner?

lchrusciel replied:
I cannot merge it with the current solution for sure. Adding a setter would be a precedent and I'm not sure if there is any major benefit of it (compared to just allowing for nulling in the constructor). Service definition will be a little bit more complicated and we still need to ensure, that this value is set at the execution time. Nonetheless, I'm not against it, I'm just challenging an idea. 

```suggestion
    public function __construct(?AvailableProductOptionValuesResolverInterface $availableProductOptionValuesResolver)
    {
        if (null === $availableProductOptionValuesResolver) {
            @trigger_error('Not passing availableProductOptionValuesResolver thru constructor is deprecated in Sylius 1.8 and it will be removed in Sylius 2.0');
        }
        
        $this->availableProductOptionValuesResolver = $availableProductOptionValuesResolver;
    }
```

mmenozzi replied:
For optional dependencies I think it's better a setter injection. Taking from symfony docs (https://symfony.com/doc/current/service_container/calls.html):

> Usually, you'll want to inject your dependencies via the constructor. But sometimes, especially if a dependency is optional, you may want to use "setter injection".

But if you don't want a precedent I'm happy with the optional constructor argument as well. Let me know, I can work on it.

lchrusciel replied:
Your point is valid, however, I would prefer to not have this argument as optional, but we need to keep it optional due to BC policy. If we will use setter injection, the upgrade path between 1.0 and 2.0 would be harder, as we would like to make it required in 2.0. With optional setter injection upgrade path would be smoother, as the same configuration will work on both versions. Let's go with optional constructor argument.

mmenozzi replied:
Done in https://github.com/Sylius/Sylius/pull/11241/commits/572fa6aa5dd8fcd660b6bf11b15188213ddfc842

mmenozzi commented on src/Sylius/Bundle/CoreBundle/EventListener/SimpleProductLockingListener.php:
```diff
@@ -46,7 +42,7 @@ public function lock(GenericEvent $event): void
 
         if ($product->isSimple()) {
             /** @var ProductVariantInterface $productVariant */
-            $productVariant = $this->variantResolver->getVariant($product);
```
Because now the implementation of `ProductVariantResolverInterface` will return the first **enabled** variant. Not the first variant.
It's kind an edge case but I thought that here, for a locking purpose, doesn't matter if the first variant is enabled or not. You only want the first variant to be locked so I removed the variant resolver in favor of `$product->getVariants()->first();`. Am I wrong?

lchrusciel replied:
Makes sense

pamil replied:
In my opinion, `ProductVariantResolver` is a service to give us some product variant that we can operate on. I guess it doesn't make much sense to lock a disabled product variant (since we can't buy it in the first place). Moreover, one might have customised their product variant resolver and this part of code assumes it's not customised.

@mmenozzi @lchrusciel what do you think?

lchrusciel replied:
I think we are all on the same page now. As a conclusion, it should be brought back, am I right?

mmenozzi replied:
Ohh ok sorry guys. Now I understand. So I'm going to revert to its original implementation.

mmenozzi replied:
Mmmm sorry guys, just one more question. Should we handle the case where there's no enabled variant for the product? And if so, how?

mmenozzi replied:
Hi @lchrusciel @pamil, I decided to proceed with the revert because given the changes in the `SimpleProductSubscriber ` here https://github.com/Sylius/Sylius/pull/11241/commits/39709bd6aa119dcf5d8761e3ea7f89d123e9eb8c, for simple products the status should be the same between product and variant (at least using the admin interface).

mmenozzi replied:
@pamil @lchrusciel, unfortunately we have failing Behat scenarios using that the variant resolver in the `SimpleProductLockingListener` (https://travis-ci.org/github/Sylius/Sylius/jobs/687327922#L3096). I think that it's because as I supposed before, while disabling or enabling a simple product, the variant is disabled and the locking listener has no variant to lock...

mmenozzi replied:
@pamil @lchrusciel I force pushed the branch without the revert because of failing tests (https://travis-ci.org/github/Sylius/Sylius/jobs/687327922#L3096).
Maybe it's because I don't understand the purpose of this SimpleProductLockingListener but I see that it works only on simple products so only on products that have only one variant. Sounds "strange" to use a variant resolver for products that have only one variant, there's no need to resolve between many variants. We should only get that variant and lock it.

Moreover, given the fact that this PR changes the variant resolver to return the first enabled variant this may result in no variant returned if it's disabled. So if we want to use the variant resolver here we should also improve this listener to support the case where there's no enabled variant. But how? And most important, why? What's the reason to do this?

pamil replied:
Indeed it makes sense for simple products to get the first variant (cause that's the only one).

mmenozzi replied:
Ok happy to hear that!
I'm going to resolve this conversation then.

lchrusciel commented:
Rebase needed as well

mmenozzi commented on features/product/viewing_products/viewing_product_enabled_variants_options_only.feature:
```diff
@@ -0,0 +1,18 @@
+@viewing_products
+Feature: Viewing product's enabled variants only
+    In order to buy only available products
+    As a Customer
+    I want to see only enabled product variants options
+
+    Background:
+        Given the store operates on a channel named "Web-US" in "USD" currency
+        And the store has a "Super Cool T-Shirt" configurable product
+        And this product has option "Size" with values "Small", "Medium" and "Large"
+        And this product has option "Color" with values "Blue", "Green" and "Yellow"
+        And this product has all possible variants
+        But all the product variants with the "Yellow" color are disabled
+
+    @ui
+    Scenario: Seeing only enabled variants options
+        When I view product "Super Cool T-Shirt"
+        Then I should not be able to select the "Yellow" color option value
```
To do what you asked in https://github.com/Sylius/Sylius/pull/11241/files#r405581869 I'd change as follows:

```suggestion
@viewing_products
Feature: Viewing product's enabled variants only
    In order to buy only available products
    As a Customer
    I want to see only enabled product variants options

    Background:
        Given the store operates on a channel named "Web-US" in "USD" currency
        And the store has a "Super Cool T-Shirt" configurable product
        And this product has option "Size" with values "Small", "Medium" and "Large"
        And this product has option "Color" with values "Blue", "Green" and "Yellow"
        And this product has all possible variants
        But all the product variants with the "Yellow" color are disabled
        And the store has a "Extra Cool T-Shirt" configurable product
        And this product has option "Size" with values "Small" and "Large"
        And this product has option "Color" with values "Yellow" and "Blue"
        And this product has all possible variants
        But the "Small" / "Blue" variant of this product is disabled
        And the "Large" / "Yellow" variant of this product is disabled

    @ui
    Scenario: Seeing only enabled variants options
        When I view product "Super Cool T-Shirt"
        Then I should not be able to select the "Yellow" color option value

    @ui
    Scenario: Seeing both values for both options when diagonal variants are available
        When I view product "Extra Cool T-Shirt"
        Then I should be able to select the "Yellow" and "Blue" color option values
        And I should be able to select the "Small" and "Large" size option values
```

What do you think? Is it ok for you?

lchrusciel replied:
I'm fine with that 👍 But we should keep the scenario with `unavailable` price as well I suppose.

mmenozzi replied:
I don't understand. Do you mean the `Viewing an unavailable message if selected option combination is unavailable` scenario in the `viewing_different_price_for_different_product_variants_selected_with_options.feature` file? If so I cannot keep it because it's not possibile to select unavailable variants anymore (as I already said here https://github.com/Sylius/Sylius/pull/11241#discussion_r405284886).

mmenozzi replied:
Ahh wait maybe I understand now. Do you want another scenario to verify that if I choose Small / Blue I get an "unavailable" message?

lchrusciel replied:
yes, that's correct

mmenozzi replied:
Done: https://github.com/Sylius/Sylius/pull/11241/commits/df192a303e3f674b113df8261cbb787633b8f637

lchrusciel commented on features/product/viewing_products/viewing_different_price_for_different_product_variants_selected_with_options.feature:
```diff
@@ -21,10 +21,4 @@ Feature: Viewing different price for different product variants selected with op
-        And I set its volume to "1L"
-        Then the product price should be "Unavailable"
-        Then I should be unable to add it to the cart
+        
```
```suggestion
```

mmenozzi commented:
Hi @lchrusciel, I integrated all your change requests. Let me know if it's ok now.
Happy easter to you and all the awesome Sylius team! 😉

lchrusciel commented:
Hey Manuele, thanks for your effort. I would really like to put it to the core. For now, I've requested my colleagues to have a second check.

What is more, we need to rebase this PR to the newest master. Can you do it?

mmenozzi commented:
Sure. Do you want rebase and squash into a single commit or just a rebase?

lchrusciel commented:
Just a rebase will be fine ;) 

And of course, happy easter to you as well!  🥚 🐇 🎉

mmenozzi commented:
Rebased @lchrusciel! 😉

mmenozzi commented:
Hi guys, what about this PR? Did you have chance to review it?

mmenozzi commented:
Hi @pamil, @Zales0123, @lchrusciel, @CoderMaggie I just rebased this PR again on top of master.
It should miss only a review because @lchrusciel already reviewed it. I hope you can find time. I really like to see this merged soon.

pamil commented on src/Sylius/Bundle/ProductBundle/Form/Type/ProductOptionValueChoiceType.php:
```diff
@@ -13,14 +13,31 @@
+    /** @var AvailableProductOptionValuesResolverInterface|null */
+    private $availableProductOptionValuesResolver;
+
+    public function __construct(?AvailableProductOptionValuesResolverInterface $availableProductOptionValuesResolver)
```
🚨 The default value for BC.

```suggestion
    public function __construct(?AvailableProductOptionValuesResolverInterface $availableProductOptionValuesResolver = null)
```

pamil commented on src/Sylius/Bundle/ProductBundle/Resources/config/services.xml:
```diff
@@ -65,5 +65,8 @@
             <tag name="doctrine.event_listener" event="postUpdate" lazy="true" />
         </service>
+
+        <service id="sylius.available_product_option_values_resolver" class="Sylius\Component\Product\Resolver\AvailableProductOptionValuesResolver" />
```
Just thinking out loud / not really related to this PR: Should we still use service names like "sylius.available_product_option_values_resolver"? Or should we go with "Sylius\Component\Product\Resolver\AvailableProductOptionValuesResolverInterface"? Don't know.

mmenozzi replied:
It's a matter of taste but I think service id notation like `sylius.available_product_option_values_resolver` is better because you can use this notation to define all the services of your application. Instead having ids like `Sylius\Component\Product\Resolver\AvailableProductOptionValuesResolverInterface` is possibile only when there's only one implementation of the interface and you need only one instance of that implementation in your application. So using FQCN as service ids you'll end up having both notations in your config files and I don't like it.

mmenozzi commented:
https://github.com/Sylius/Sylius/pull/11241#discussion_r425077562

@pamil I agree indeed the `DefaultProductVariantResolver` has been modified in this PR to return the first **enabled** variant.

keeper7 commented:
Assume that I have enabled product with two variants. What will happend if I disable both variants?
I think there should be listener to either
1. disable product when no enabled variants exists
2. or prevent disabling last variant.
Also when you want to enable product with zero enabled variants it should prevent it.

mmenozzi commented:
It could make sense but I think it could be done even in a separate PR. @pamil and @lchrusciel what do you think?

lchrusciel commented:
Hey folks! 

In default Sylius installation you are able to list products without any variant configured. It is a non-buyable product(where there is out-of-stock information displayed) and I would expect the same behavior after all variants were toggled off.

mmenozzi commented:
Ok so no listener to disable product when all variants are disabled.

pamil commented:
Same bug when accessing the enabled, configurable product page with all variants disabled.

mmenozzi commented:
Thank you @pamil I'll cover all those cases with Behat scenarios to make sure to not introduce regressions in the future.

mmenozzi commented:
Ok I'll add a Behat scenario for that also.

mmenozzi commented:
Hi @pamil I added the Behat scenarios and fixed the code for:
* Viewing enabled product with all variants disabled from taxon page
* Viewing enabled product with all variants disabled from product show page
* Checking out with a disabled product variant

pamil commented:
Thank you, Manuele! :1st_place_medal:

mmenozzi commented:
🎉🎉🎉
Thank you for your awesome work guys!

pamil commented:
It's time for #11287 now :)

mmenozzi commented:
Toggleable all the things!

loevgaard commented:
Hi @mmenozzi. Just saw this PR. Thank you very much <3

mmenozzi commented:
> Hi @mmenozzi. Just saw this PR. Thank you very much <3

You're welcome! You'll have to take into consideration for the https://github.com/Setono/SyliusFeedPlugin (and probably others) 😉.
